src/app/digital/utils/ExpressionParser/GenerateInputTree.ts

```diff
@@ -0,0 +1,144 @@
+ * @see GenerateInputTree
+ */
+ function generateInputTreeCore(tokens: Array<Token>, ops: Map<FormatLabels, string>, currentOp: TokenType = "|", index: number = 0): NewTreeRetValue {
+    const nextOp = DefaultPrecedences.get(currentOp);
+    if (tokens[index].type === ")") {
+        if (index > 0) {
+            if (tokens[index-1].type === "(")
+                throw new Error("Empty Parenthesis");
+            else if (IsOperator(tokens[index-1]))
+                throw new Error("Missing Right Operand: \"" + ops.get(tokens[index-1].type as FormatLabels) + "\"");
+        }
+        throw new Error("Encountered Unmatched \"" + ops.get(")") + "\"");
+    }
+
+    // When this function has recursed through to "!" and the token still isn't that, then
+    //  the only possibilites left are an input or open parenthesis token
+    if (currentOp === "!" && tokens[index].type !== "!") {
+        const token = tokens[index];
+        if (token.type === "(")
+            return generateInputTreeCore(tokens, ops, nextOp, index);
+        else if (token.type === "input")
+            return {index: index+1, tree: {kind: "leaf", ident: token.name}};
+        else
+            throw new Error("Missing Left Operand: \"" + ops.get(token.type) + "\"");
+    }
+
+    // This section gets the part of the tree from the left side of the operator.
+    //  "!" and "(" only have operands on their right side, so this section is skipped for them
+    let leftRet: NewTreeRetValue;
+    if (currentOp !== "!" && currentOp !== "(") {
+        leftRet = generateInputTreeCore(tokens, ops, nextOp, index);
+        index = leftRet.index;
+        // If this isn't the right operation to apply, return
+        if (index >= tokens.length || tokens[index].type !== currentOp)
+            return leftRet;
+    }
+
+    // This section gets the part of the tree from the right side of the operand. index is incremented by 1
+    //  so it now points to the token on the right side of the operator.
+    index += 1;
+    if (index >= tokens.length && currentOp !== "(") {
+        throw new Error("Missing Right Operand: \"" + ops.get(currentOp) + "\"");
+    }
+    let rightRet: NewTreeRetValue = null;
+    const rightToken = tokens[index];
+    if (currentOp === "!" && rightToken.type === "!") // This case applies when there are two !'s in a row
+        rightRet = generateInputTreeCore(tokens, ops, currentOp, index);
+    else if (currentOp === "!" && rightToken.type === "input") // This case would apply when an input follows a "!"
+        rightRet = {index: index+1, tree: {kind: "leaf", ident: rightToken.name}};
+    else if (currentOp === "(") {
+        if (index >= tokens.length)
+            throw new Error("Encountered Unmatched \"" + ops.get("(") + "\"");
+        rightRet = generateInputTreeCore(tokens, ops, nextOp, index);
+    }
+    else
+        rightRet = generateInputTreeCore(tokens, ops, currentOp, index);
```

Comment:
nit: Don't mix braces and non-braces, also can keep elseif/elses on same line as ending `}`